File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {    
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        function<TreeNode*(TreeNode*)> findAncestor = [&](TreeNode* root) -> TreeNode* {
            if (!root) return nullptr;

            if (p->val < root->val && q->val < root->val) return findAncestor(root->left);
            else if (p->val > root->val && q->val > root->val) return findAncestor(root->right);
            else return root;
        };

        return findAncestor(root);
    }
};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-4.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        if (preorder.Length == 0 || inorder.Length == 0) return null;

        TreeNode root = new TreeNode(preorder[0]);
        int mid = Array.IndexOf(inorder, preorder[0]);

        root.left = BuildTree(preorder[1..(mid + 1)], inorder[..mid]);
        root.right = BuildTree(preorder[(mid + 1)..], inorder[(mid + 1)..]);

        return root;
    }
}
